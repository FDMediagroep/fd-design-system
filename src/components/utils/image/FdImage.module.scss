.image {
    min-height: 100%; // The image can grow if the content grows
    object-fit: cover;
}
