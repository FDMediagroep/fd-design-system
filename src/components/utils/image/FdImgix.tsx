import React, { JSX } from 'react';
import Imgix, { SharedImgixAndSourceProps } from 'react-imgix';

import styles from './FdImage.module.scss';

type cssValue = `${number}${'px' | 'rem' | 'vw'}`;

interface Sizes {
    /**
     * How many columns wide is the image of this component on S screens
     */
    widthS: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | cssValue;
    /**
     * How many columns wide is the image of this component on M screens
     * @default defaults to the value of colS
     */
    widthM?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | cssValue;
    /**
     * How many columns wide is the image of this component on L screens
     * @default defaults to the value of colS
     */
    widthL?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | cssValue;
}

interface Crop {
    ratio: string;
    x: number;
    y: number;
    width: number;
    height: number;
}

interface Props extends Omit<SharedImgixAndSourceProps, 'sizes' | 'src'> {
    imageSizes: Sizes;
    crop?: Crop;
    src?: SharedImgixAndSourceProps['src'];
}

export const noImagePlaceholder = `${process.env.NEXT_PUBLIC_IMGIX_BASE_URL}/2b57bde2582fa63dc980a53f1a021ac5177ca468.png`;

/**
 * This component sets the correct sizes to load images based on our grid
 */
export function FdImgix({
    imageSizes,
    crop,
    className = '',
    ...imgixProps
}: Props): JSX.Element {
    const imageSizesWidth: Array<string> = [];

    // Calculate sizes for L screens
    if (imageSizes.widthL) {
        let lSize = imageSizes.widthL.toString();
        if (Number.isInteger(imageSizes.widthL)) {
            lSize = `calc(1080px / 12 * ${imageSizes.widthL})`;
        }
        imageSizesWidth.push(`(min-width: 1080px) ${lSize}`);
    }

    // Calculate sizes for M screens
    if (imageSizes.widthM) {
        let mSize = imageSizes.widthM.toString();
        if (Number.isInteger(imageSizes.widthM)) {
            mSize = `calc(768px / 12 * ${imageSizes.widthM})`;
        }
        imageSizesWidth.push(`(min-width: 768px) ${mSize}`);
    }

    // Calculate sizes for S screens
    let sSize = imageSizes.widthS.toString();
    if (Number.isInteger(imageSizes.widthS)) {
        sSize = `calc(100vw / 12 * ${imageSizes.widthS})`;
    }
    imageSizesWidth.push(sSize);

    if (crop && imgixProps.src) {
        imgixProps.imgixParams = imgixProps.imgixParams || {};
        imgixProps.imgixParams.rect = [crop.x, crop.y, crop.width, crop.height]
            .map((cropValue) => {
                /**
                 * Crops only work if the value starts with a '.'
                 * So if we got a number value we need to cast to a string and remove the first 0
                 */
                const stringValue = cropValue.toString();
                if (stringValue.startsWith('0.')) {
                    return stringValue.substring(1);
                } else if (stringValue === '1') {
                    return `${stringValue}.0`;
                }
                return stringValue;
            })
            .join(',');
    }

    // default imgix parameters
    if (imgixProps.imgixParams) {
        imgixProps.imgixParams.q = imgixProps.imgixParams.q ?? '45';
    }

    return (
        <>
            {!imgixProps.src && (
                <Imgix
                    className={`${styles['image']} ${className}`}
                    sizes={imageSizesWidth.join(', ')}
                    {...(imgixProps as SharedImgixAndSourceProps)}
                    src={`${noImagePlaceholder}`}
                />
            )}
            {imgixProps.src && (
                <Imgix
                    className={`${styles['image']} ${className}`}
                    sizes={imageSizesWidth.join(', ')}
                    {...(imgixProps as SharedImgixAndSourceProps)}
                />
            )}
        </>
    );
}
