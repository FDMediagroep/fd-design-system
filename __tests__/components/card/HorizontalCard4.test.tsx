import React from 'react';
import {
    HorizontalCard4,
    getCssClassNames,
} from '../../../src/components/card/HorizontalCard4';
import { render } from '@testing-library/react';
import { FdImgix } from '../../../src/components/utils/image/FdImgix';

describe('HorizontalCard4', () => {
    test('should render correctly', () => {
        const { container, getByText } = render(
            <HorizontalCard4
                id="1"
                url="https://example.com/1/This is the title"
                label="Ondernemen"
                time="11:40"
                title="This is the title"
                isRead={false}
                imageComponent={
                    <FdImgix
                        src="https://images.example.com/image.png"
                        imageSizes={{
                            widthS: '160px',
                            widthM: '244px',
                            widthL: '244px',
                        }}
                    />
                }
            />
        );

        expect(getByText('This is the title')).toBeTruthy();
        expect(getByText('Ondernemen')).toBeTruthy();
        expect(getByText('11:40')).toBeTruthy();
        expect(container?.querySelector('[href]')?.getAttribute('href')).toBe(
            'https://example.com/1/This is the title'
        );
    });

    test('should render image correctly', () => {
        const { container } = render(
            <HorizontalCard4
                id="1"
                url="https://example.com/1/This is the title"
                label="Ondernemen"
                time="11:40"
                title="This is the title"
                isRead={false}
                imageComponent={
                    <FdImgix
                        src="https://images.example.com/image.png"
                        imageSizes={{
                            widthS: '160px',
                            widthM: '244px',
                            widthL: '244px',
                        }}
                    />
                }
            />
        );
        expect(container?.querySelector('img')?.getAttribute('src')).toBe(
            'https://images.example.com/image.png?auto=format&ixlib=react-9.5.1-beta.1'
        );
    });

    test('should render is-read correctly', () => {
        const { container } = render(
            <HorizontalCard4
                id="1"
                url="https://example.com/1/This is the title"
                label="Ondernemen"
                time="11:40"
                title="This is the title"
                isRead={true}
                imageComponent={
                    <FdImgix
                        src="https://images.example.com/image.png"
                        imageSizes={{
                            widthS: '160px',
                            widthM: '244px',
                            widthL: '244px',
                        }}
                    />
                }
            />
        );
        expect(container).toMatchSnapshot();
    });

    test('should have correct class name', () => {
        const { container } = render(
            <HorizontalCard4
                id="1"
                url="https://example.com/1/This is the title"
                label="Ondernemen"
                time="11:40"
                title="This is the title"
                isRead={false}
                imageComponent={
                    <FdImgix
                        src="https://images.example.com/image.png"
                        imageSizes={{
                            widthS: '160px',
                            widthM: '244px',
                            widthL: '244px',
                        }}
                    />
                }
            />
        );

        expect(
            container?.firstElementChild?.classList.contains(
                getCssClassNames()[0]
            )
        ).toBeTruthy();
    });

    test('should render without image component', () => {
        const { getByRole } = render(
            <HorizontalCard4
                id="1"
                url="https://example.com/1/This is the title"
                label="Ondernemen"
                time="11:40"
                title="This is the title"
                isRead={false}
            />
        );
        const image = getByRole('img');
        expect(image).toHaveProperty(
            'src',
            expect.stringContaining('/assets/images/fd-logo.svg')
        );
    });
});
