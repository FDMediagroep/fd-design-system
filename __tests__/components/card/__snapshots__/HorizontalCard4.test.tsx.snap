// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalCard4 should render is-read correctly 1`] = `
<div>
  <article
    class="fdHorizontalCard4 isRead xs__p+4"
    id="1"
  >
    <a
      aria-label="This is the title"
      href="https://example.com/1/This is the title"
    >
      <figure
        class="xs__m-0"
      >
        <img
          class="image "
          sizes="(min-width: 1080px) 244px, (min-width: 768px) 244px, 160px"
          src="https://images.example.com/image.png?auto=format&ixlib=react-9.8.1"
          srcset="https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=100 100w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=116 116w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=135 135w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=156 156w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=181 181w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=210 210w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=244 244w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=283 283w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=328 328w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=380 380w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=441 441w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=512 512w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=594 594w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=689 689w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=799 799w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=927 927w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=1075 1075w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=1247 1247w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=1446 1446w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=1678 1678w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=1946 1946w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=2257 2257w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=2619 2619w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=3038 3038w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=3524 3524w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=4087 4087w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=4741 4741w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=5500 5500w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=6380 6380w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=7401 7401w,
https://images.example.com/image.png?auto=format&ixlib=react-9.8.1&w=8192 8192w"
        />
        <figcaption />
      </figure>
      <div
        class="teaserText xs__pl+4 s__p+4"
      >
        <div
          class="meta xs__m-0 xs__mb+2 body-text sans xs"
        >
          <time>
            11:40
          </time>
          <span
            class="dot"
          >
            â€¢
          </span>
          <span
            class="prefix"
          >
            Ondernemen
          </span>
        </div>
        <div
          class="textContainer"
        >
          <h1
            class="heading serif s xs__m-0"
          >
            This is the title
          </h1>
        </div>
      </div>
    </a>
  </article>
</div>
`;
